cinema/views.py

```diff
@@ -32,27 +41,91 @@ class CinemaHallViewSet(viewsets.ModelViewSet):
 
 class MovieViewSet(viewsets.ModelViewSet):
     queryset = Movie.objects.all()
-    serializer_class = MovieSerializer
 
     def get_serializer_class(self):
         if self.action == "list":
             return MovieListSerializer
-
         if self.action == "retrieve":
             return MovieDetailSerializer
-
         return MovieSerializer
 
+    def get_queryset(self):
+        queryset = self.queryset.prefetch_related("genres", "actors")
+
+        if self.action == "list":
+            actors = self.request.query_params.get("actors")
+            genres = self.request.query_params.get("genres")
+            title = self.request.query_params.get("title")
+
+            if actors:
+                actor_ids = list(map(int, actors.split(",")))
+                queryset = queryset.filter(actors__id__in=actor_ids)
+
+            if genres:
+                genre_ids = list(map(int, genres.split(",")))
+                queryset = queryset.filter(genres__id__in=genre_ids)
+
+            if title:
+                queryset = queryset.filter(title__icontains=title)
+
+        return queryset.distinct()
+
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
-    serializer_class = MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = self.queryset.select_related(
+            "movie", "cinema_hall"
+        ).prefetch_related("tickets__order")
+
+        if self.action == "list":
+            movie = self.request.query_params.get("movie")
+            date = self.request.query_params.get("date")
+
+            if movie:
+                queryset = queryset.filter(movie_id=movie)
+
+            if date:
+                queryset = queryset.filter(show_time__date=date)
+
+            queryset = queryset.annotate(
+                tickets_available=(
+                    F("cinema_hall__rows") * F("cinema_hall__seats_in_row")
+                ) - Count("tickets")
+            )
+
+        return queryset.distinct()
 
     def get_serializer_class(self):
         if self.action == "list":
             return MovieSessionListSerializer
-
         if self.action == "retrieve":
             return MovieSessionDetailSerializer
-
         return MovieSessionSerializer
+
+
+class OrderSetPagination(PageNumberPagination):
+    page_size = 8
+    page_size_query_param = "page_size"
+    max_page_size = 20
+
+
+class OrderViewSet(viewsets.ModelViewSet):
+    queryset = Order.objects.all()
+    serializer_class = OrderSerializer
+    pagination_class = OrderSetPagination
+
```

Comment:
get_serializer_class will need to be implemented here after you add another OrderSerializer